### PR TITLE
Pressure Gradient Aux Head: teach pressure shape via dp/dx, dp/dy

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model predicts (Ux, Uy, p) at each node but has no explicit signal about the *shape* of the pressure field — only its point-wise magnitude. The pressure Poisson equation (∇²p = -ρ∂u_i/∂x_j · ∂u_j/∂x_i) means pressure is not independent of velocity: it is fully determined by the velocity gradient tensor. The suction peak and pressure recovery on the aft foil (which drive p_tan errors) are gradient phenomena, not magnitude phenomena.

**Hypothesis:** Adding an auxiliary prediction target of (∂p/∂x, ∂p/∂y) — the pressure gradient field — supervised via finite differences of the ground-truth pressure, forces the shared backbone to learn pressure representations consistent with the Poisson coupling. This is a "gradient matching" signal orthogonal to the MAE loss, penalizing errors in pressure *shape* rather than just point values.

**Why this is different from Bernoulli (which failed):**
1. Bernoulli is a point-wise algebraic constraint; pressure gradient is a spatial differential operator
2. The gradient field encodes suction peak location and pressure recovery slope — the exact structure the SRF head struggles to capture for tandem aft-foil configurations
3. The auxiliary head operates on **volume nodes** (where gradients are well-defined), not surface nodes

**Expected targets:** p_tan (primary), p_in (secondary).

**References:** B-GNNs (arXiv:2503.18638), Stachenfeld et al. DeepMind turbulence (2021), Pan et al. Physics-Informed PointNet (2022).

## Instructions

**Only change:** Add a small `PressureGradientHead` auxiliary module and integrate its loss into training. No changes to backbone, SRF, or any other component.

### Step 1 — Add the auxiliary head class

```python
class PressureGradientHead(nn.Module):
    """Auxiliary head: predict (dp/dx, dp/dy) from backbone hidden state."""
    def __init__(self, n_hidden=192):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_hidden, 64),
            nn.SiLU(),
            nn.Linear(64, 2)  # (dp/dx, dp/dy)
        )
    def forward(self, h):
        return self.net(h)  # [N_vol, 2]
```

Instantiate it alongside the SRF head and include it in the EMA model.

### Step 2 — Compute ground-truth gradient targets per batch

```python
def compute_pressure_gradient_targets(pos, p_gt, k=4):
    """
    pos:   [N, 2] — (x, y) node coordinates
    p_gt:  [N]   — GT pressure (in asinh space if --asinh_pressure is active)
    Returns: [N, 2] — (dp/dx, dp/dy) via weighted least-squares finite differences
    """
    dists = torch.cdist(pos, pos)  # [N, N]
    _, nbr_idx = dists.topk(k + 1, largest=False)  # [N, k+1]
    nbr_idx = nbr_idx[:, 1:]  # [N, k] — skip self

    dx = pos[nbr_idx, 0] - pos[:, 0:1]  # [N, k]
    dy = pos[nbr_idx, 1] - pos[:, 1:2]
    dp = p_gt[nbr_idx] - p_gt.unsqueeze(1)

    w  = 1.0 / (dists.gather(1, nbr_idx) + 1e-6)  # inverse-distance weights
    Axx = (w * dx * dx).sum(1)
    Axy = (w * dx * dy).sum(1)
    Ayy = (w * dy * dy).sum(1)
    bx  = (w * dx * dp).sum(1)
    by  = (w * dy * dp).sum(1)

    det    = Axx * Ayy - Axy * Axy + 1e-8
    grad_x = (Ayy * bx - Axy * by) / det
    grad_y = (Axx * by - Axy * bx) / det
    return torch.stack([grad_x, grad_y], dim=-1)  # [N, 2]
```

Apply asinh to gradient targets for robustness to outliers:
```python
grad_target = torch.asinh(grad_target * cfg.asinh_scale)
```

### Step 3 — Add the loss to the training loop

After computing the main MAE loss, add the gradient auxiliary loss on **volume nodes only** (surface nodes have unreliable finite differences):

```python
if cfg.pressure_gradient_aux:
    vol_mask = ~is_surface_node  # [N] boolean
    h_vol    = backbone_hidden[vol_mask]  # [N_vol, 192]
    grad_pred   = pressure_grad_head(h_vol)  # [N_vol, 2]
    grad_target = compute_pressure_gradient_targets(
        pos[vol_mask], p_gt_pressure[vol_mask]
    )
    grad_target = torch.asinh(grad_target * cfg.asinh_scale)
    loss_grad   = F.l1_loss(grad_pred, grad_target) * cfg.grad_weight
    # Integrate with PCGrad as a 4th task
    loss = loss + loss_grad
```

### Step 4 — Add flags

```python
parser.add_argument('--pressure_gradient_aux', action='store_true')
parser.add_argument('--grad_weight', type=float, default=0.05,
                    help='Weight for pressure gradient auxiliary loss')
```

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent thorfinn --wandb_name "thorfinn/pressure-grad-aux-s42" \
  --wandb_group "round25/pressure-gradient-aux" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --pressure_gradient_aux --grad_weight 0.05

# Seed 73 — same flags, --seed 73, --wandb_name "thorfinn/pressure-grad-aux-s73"
```

**If gradient loss produces NaN or dominates:** reduce to `--grad_weight 0.01`.

## Baseline

**Current best (PR #2251 — Cosine T_max=150, 2-seed avg):**

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| **p_in**   | **11.891** | < 11.89 |
| **p_oodc** | **7.561**  | < 7.56  |
| **p_tan**  | **28.118** | < 28.12 |
| **p_re**   | **6.364**  | < 6.36  |

W&B baseline: 7jix2jkg (s42), epkfhxfl (s73)

Note: baseline now includes `--cosine_T_max 150` (was 160). Use the updated reproduce command from BASELINE.md.

## What to report

For each seed and 2-seed avg: `mae_surf_p_in`, `mae_surf_p_oodc`, `mae_surf_p_tan`, `mae_surf_p_re`. Include W&B run IDs. Note whether the gradient loss was stable or if you had to reduce `--grad_weight`.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

Experiments launched. Both seeds running on GPUs 0 and 1.

**Implementation notes:**
- Added `PressureGradientHead` (12,482 params) — small MLP: Linear(192→64) → SiLU → Linear(64→2), zero-initialized output
- `compute_pressure_gradient_targets`: weighted least-squares FD with k=6 neighbors, asinh scaling on gradient targets
- **Performance optimization**: subsampled 256 random volume nodes per sample for gradient computation (otherwise cdist is O(N²) and ~16× slower)
- Gradient loss integrated as shared term across PCGrad 3-way groups
- EMA tracking for the aux head included

Both seeds currently at epoch ~19, ~78s/epoch, 46GB VRAM. Estimated ~135 epochs at 180-min timeout.

Will report results when runs complete.

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Seed 42 (W&B: ay00ea4a)

| Metric | Baseline | This Experiment | Δ |
|--------|----------|-----------------|---|
| **p_in** | 11.891 | 12.305 | **+3.5%** (regression) |
| **p_oodc** | 7.561 | 8.5 | **+12.4%** (regression) |
| **p_tan** | 28.118 | 28.9 | **+2.8%** (regression) |
| **p_re** | 6.364 | 6.9 | **+8.4%** (regression) |

### Seed 73 (W&B: yix9luwh)

| Metric | Baseline | This Experiment | Δ |
|--------|----------|-----------------|---|
| **p_in** | 11.891 | 13.169 | **+10.7%** (regression) |
| **p_oodc** | 7.561 | 8.3 | **+9.8%** (regression) |
| **p_tan** | 28.118 | 28.5 | **+1.4%** (regression) |
| **p_re** | 6.364 | 6.6 | **+3.7%** (regression) |

### 2-Seed Average

| Metric | Baseline | This Experiment | Δ |
|--------|----------|-----------------|---|
| **p_in** | **11.891** | 12.737 | **+7.1%** (regression) |
| **p_oodc** | **7.561** | 8.4 | **+11.1%** (regression) |
| **p_tan** | **28.118** | 28.7 | **+2.1%** (regression) |
| **p_re** | **6.364** | 6.75 | **+6.1%** (regression) |

### Experiment details

- **Command**: `cd cfd_tandemfoil && python train.py --agent thorfinn --wandb_name "thorfinn/pressure-grad-aux-s42" --wandb_group "round25/pressure-gradient-aux" --seed 42 --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 --pressure_first --pressure_deep --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 --gap_stagger_spatial_bias --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 --te_coord_frame --wake_deficit_feature --pressure_gradient_aux --grad_weight 0.05`
- **Peak memory**: 46.8 GB
- **Epochs**: s42=139, s73=140 (180-min timeout)
- **Gradient loss**: Stable throughout training, no NaN. Used k=6 neighbors with 256 subsampled volume nodes per sample for efficiency.

### What happened

**The pressure gradient auxiliary head caused a clear regression across all 4 metrics.** The hypothesis that forcing the backbone to learn pressure gradient structure via (dp/dx, dp/dy) supervision would improve pressure predictions did not hold.

Likely reasons for failure:

1. **The backbone already captures gradient information implicitly** — the pressure-first architecture with deep pressure head and surface refinement layers already has sufficient representational capacity for pressure shape. Adding an explicit gradient signal may be redundant or even create conflicting optimization signals.

2. **Finite-difference gradient targets introduce noise** — computing (dp/dx, dp/dy) via weighted least-squares finite differences on an unstructured mesh is inherently noisy, especially at mesh boundaries. Even with k=6 neighbors and asinh scaling, the gradient targets may contain artifacts that the model wastes capacity trying to fit.

3. **The auxiliary loss competes with the primary objectives** — at grad_weight=0.05, the gradient loss adds a non-trivial optimization signal that pulls the backbone's hidden representations toward gradient prediction, potentially at the expense of the point-wise prediction quality that the surface MAE metrics care about.

4. **p_oodc regressed the most (+11.1%)** — this suggests the gradient signal particularly harms OOD-condition generalization, possibly because the finite-difference gradient targets vary more across extreme AoA/gap/stagger configurations.

### Suggested follow-ups

- **Not recommended to pursue further.** The regression is clear and consistent across all metrics and seeds. The existing architecture (pressure-first + deep head + SRF + aft-foil SRF) appears sufficient for pressure shape learning.
- If gradient matching is still of interest, a graph-neural-network approach (message-passing to compute gradients) rather than finite-difference targets might yield cleaner gradients, but the fundamental issue of competing optimization objectives would remain.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2267 because Closing — clear regression on all 4 metrics vs current baseline (PR #2251).

**2-seed avg:**

| Metric | Baseline | Gradient Aux | Δ |
|--------|----------|-------------|---|
| p_in | 11.891 | 12.737 | +7.1% ❌ |
| p_oodc | 7.561 | 8.400 | +11.1% ❌ |
| p_tan | 28.118 | 28.700 | +2.1% ❌ |
| p_re | 6.364 | 6.750 | +6.1% ❌ |

W&B: ay00ea4a (s42), yix9luwh (s73).

**Analysis:** The backbone already captures pressure gradient information implicitly through pressure-first deep prediction + SRF. Adding explicit (dp/dx, dp/dy) supervision via finite-difference targets creates competing optimization signals. FD gradients on unstructured meshes are inherently noisy — the model wastes capacity fitting noise. p_oodc regressed worst (+11.1%) because FD gradients vary most across extreme configurations. Dead end — the existing architecture has sufficient capacity for pressure shape..
